File: web/types/terminal.ts
import type { LaunchProviderSelection } from "./launch-profile";

/**
 * 标签与终端相关类型定义
 */

/** CLI 工具类型（已知值自动补全 + 允许任意字符串） */
export type KnownCliTool =
  | "none"
  | "claude"
  | "codex"
  | "gemini"
  | "kimi"
  | "glm"
  | "opencode"
  | "cursor";
export type CliTool = KnownCliTool | (string & {});

/** CLI 工具元信息（来自 Rust cc-cli-adapters crate） */
export interface CliToolInfo {
  id: string;
  displayName: string;
  executable: string;
  installed: boolean;
  version: string | null;
  path: string | null;
  capabilities?: CliToolCapabilities | null;
}

/** CLI 工具能力声明 */
export interface CliToolCapabilities {
  supportsProvider: boolean;
  supportsResume: boolean;
  supportsMcp: boolean;
  supportsSystemPrompt: boolean;
  supportsWorkspace: boolean;
  supportsProjectHooks: boolean;
  compatibleProviderTypes: string[];
}

/** WSL 启动信息 */
export interface WslLaunchInfo {
  remotePath: string;
  distro?: string;
}

export type TerminalPaneNode = TerminalPaneLeaf | TerminalPaneSplit;

export interface TerminalPaneLeaf {
  type: "leaf";
  id: string;
  /** Live PTY session id owned by CC-Panes. */
  sessionId: string | null;
  /** Agent conversation resume id, e.g. Claude/Codex resume UUID. */
  resumeId?: string;
  workspaceName?: string;
  providerId?: string;
  providerSelection?: LaunchProviderSelection;
  launchProfileId?: string;
  workspacePath?: string;
  workspaceSnapshotId?: string;
  launchClaude?: boolean;
  cliTool?: CliTool;
  ssh?: import("./workspace").SshConnectionInfo;
  wsl?: WslLaunchInfo;
  machineName?: string;
  disconnected?: boolean;
  restoring?: boolean;
  savedSessionId?: string;
}

export interface TerminalPaneSplit {
  type: "split";
  id: string;
  direction: "horizontal" | "vertical";
  children: TerminalPaneNode[];
  sizes: number[];
}

/** 通用标签 */
export interface Tab {
  id: string;
  title: string;
  contentType: "terminal" | "mcp-config" | "skill-manager" | "memory-manager" | "file-explorer" | "editor";
  projectId: string;
  projectPath: string;
  /** Live PTY session id owned by CC-Panes. */
  sessionId: string | null;
  pinned?: boolean;
  minimized?: boolean;
  /** Agent conversation resume id, e.g. Claude/Codex resume UUID. */
  resumeId?: string;
  workspaceName?: string;
  providerId?: string;
  providerSelection?: LaunchProviderSelection;
  launchProfileId?: string;
  workspacePath?: string;
  workspaceSnapshotId?: string;
  launchClaude?: boolean;
  cliTool?: CliTool;
  filePath?: string;
  dirty?: boolean;
  reclaimKey?: number;
  ssh?: import("./workspace").SshConnectionInfo;
  wsl?: WslLaunchInfo;
  machineName?: string;
  disconnected?: boolean;
  restoring?: boolean;
  savedSessionId?: string;
  terminalRootPane?: TerminalPaneNode;
  activeTerminalPaneId?: string;
  /**
   * Parent tab id when this tab was created by `launch_task` from another
   * cc-panes-managed Claude instance. Drives hierarchical numbering
   * (`#N.M`, `#N.M.K`). Top-level tabs leave it unset.
   */
  parentTabId?: string;
}

/** 终端会话状态 */
export interface TerminalSession {
  id: string;
  projectPath: string;
  cols: number;
  rows: number;
  running: boolean;
}

/** 创建终端会话请求 */
export interface CreateSessionRequest {
  launchId?: string;
  projectPath: string;
  cols: number;
  rows: number;
  workspaceName?: string;
  providerId?: string;
  providerSelection?: LaunchProviderSelection;
  launchProfileId?: string;
  workspacePath?: string;
  workspaceSnapshotId?: string;
  launchClaude?: boolean;
  cliTool?: CliTool;
  resumeId?: string;
  skipMcp?: boolean;
  appendSystemPrompt?: string;
  ssh?: import("./workspace").SshConnectionInfo;
  wsl?: WslLaunchInfo;
}

/** 打开终端的选项 */
export interface OpenTerminalOptions {
  path: string;
  workspaceName?: string;
  providerId?: string;
  providerSelection?: LaunchProviderSelection;
  launchProfileId?: string;
  workspacePath?: string;
  workspaceSnapshotId?: string;
  cliTool?: CliTool;
  resumeId?: string;
  ssh?: import("./workspace").SshConnectionInfo;
  wsl?: WslLaunchInfo;
  machineName?: string;
}

/** 终端输出事件 */
export interface TerminalOutput {
  sessionId: string;
  data: string;
}

/** 终端调整大小请求 */
export interface ResizeRequest {
  sessionId: string;
  cols: number;
  rows: number;
}

File: cc-panes-core/src/models/terminal.rs
use serde::{Deserialize, Serialize};

use crate::models::launch_profile::LaunchProviderSelection;

/// CLI 工具类型
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum CliTool {
    #[default]
    None,
    Claude,
    Codex,
    Gemini,
    Kimi,
    Glm,
    Opencode,
    Cursor,
}

impl CliTool {
    /// 转换为 CLI 适配器注册表中的 id 字符串
    pub fn as_id(&self) -> &str {
        match self {
            CliTool::None => "none",
            CliTool::Claude => "claude",
            CliTool::Codex => "codex",
            CliTool::Gemini => "gemini",
            CliTool::Kimi => "kimi",
            CliTool::Glm => "glm",
            CliTool::Opencode => "opencode",
            CliTool::Cursor => "cursor",
        }
    }
}

/// WSL 启动信息
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WslLaunchInfo {
    pub remote_path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub workspace_remote_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub distro: Option<String>,
}

/// 创建终端会话请求
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub launch_id: Option<String>,
    pub project_path: String,
    pub cols: u16,
    pub rows: u16,
    pub workspace_name: Option<String>,
    pub provider_id: Option<String>,
    #[serde(default)]
    pub provider_selection: LaunchProviderSelection,
    pub launch_profile_id: Option<String>,
    pub workspace_path: Option<String>,
    pub workspace_snapshot_id: Option<String>,
    #[serde(default)]
    pub launch_claude: bool,
    #[serde(default)]
    pub cli_tool: CliTool,
    pub resume_id: Option<String>,
    #[serde(default)]
    pub skip_mcp: bool,
    pub append_system_prompt: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh: Option<crate::models::workspace::SshConnectionInfo>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wsl: Option<WslLaunchInfo>,
}

impl CreateSessionRequest {
    /// 兼容映射：优先使用 cli_tool，fallback 到 launch_claude
    pub fn effective_cli_tool(&self) -> CliTool {
        match self.cli_tool {
            CliTool::None if self.launch_claude => CliTool::Claude,
            other => other,
        }
    }
}

/// 调整终端大小请求
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResizeRequest {
    pub session_id: String,
    pub cols: u16,
    pub rows: u16,
}

/// 终端输出事件
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutput {
    pub session_id: String,
    pub data: String,
}

/// 终端重放快照的屏幕模式
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TerminalBufferMode {
    Normal,
    Alternate,
}

/// attach-existing 时用于重建当前屏幕状态的原始 VT 快照
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalReplaySnapshot {
    pub data: String,
    pub buffer_mode: TerminalBufferMode,
}

/// 终端退出事件
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalExit {
    pub session_id: String,
    pub exit_code: i32,
}
